### PR TITLE
Don't report modifier errors while parsing.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Note: this PR depends on https://github.com/dotnet/roslyn/pull/16338 which shoudl be reviewed first.

**Comment by CyrusNajmabadi** [comment]
I also recommend reviewing https://github.com/dotnet/roslyn/pull/16341 first.

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 1]
```diff
@@ -40,17 +40,19 @@ internal sealed class LocalFunctionSymbol : MethodSymbol
             _syntax = syntax;
             _containingSymbol = containingSymbol;
 
+            var diagnostics = DiagnosticBag.GetInstance();
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: I think this new variable is unused?

**Comment by CyrusNajmabadi** [comment]
@ManishJayaswal 

Customer scenario

None. This is general infrastructure work for the compiler based on design changes the team agreed on.

Bugs this fixes:

N/A

Workarounds, if any

N/A

Risk

Low.

Performance impact

Low.

Is this a regression from a previous update?

N/A

Root cause analysis:

N/A. 

How was the bug found?

N/A

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: You are correct.  Removing.

**Comment by CyrusNajmabadi** [comment]
@ManishJayaswal 

Customer scenario

None. This is general infrastructure work for the compiler based on design changes the team agreed on.

Bugs this fixes:

N/A

Workarounds, if any

N/A

Risk

Low.

Performance impact

Low.

Is this a regression from a previous update?

N/A

Root cause analysis:

N/A. 

How was the bug found?

N/A

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGertz Improvements in parser that benefit IDE and compiler.

**Comment by MattGertz** [comment]
Approved pending tests.

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
Rebased to 15.6

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi This branch is fairly stale, should it still be active?

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski Yes it should.  I'm still waiting on a second compiler  review.  @jcouv could  you look when you have a chance?

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 2]
```diff
@@ -529,7 +529,7 @@ internal enum ErrorCode
         ERR_PartialMethodInvalidModifier = 750,
         ERR_PartialMethodOnlyInPartialClass = 751,
         ERR_PartialMethodCannotHaveOutParameters = 752,
-        ERR_PartialMethodOnlyMethods = 753,
+        // ERR_PartialMethodOnlyMethods = 753, Removed as it is subsumed by ERR_PartialMisplaced
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: >ERR_PartialMethodOnlyMethods [](start = 11, length = 28)

Did you remove the corresponding resource string?

File: src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs [hunk 3]
```diff
@@ -154,13 +156,27 @@ private static DeclarationModifiers ToDeclarationModifier(SyntaxKind kind)
             }
         }
 
-        public static DeclarationModifiers ToDeclarationModifiers(this SyntaxTokenList modifiers)
+        public static DeclarationModifiers ToDeclarationModifiers(
+            this SyntaxTokenList modifiers, bool allowPartial, DiagnosticBag diagnostics)
         {
             var result = DeclarationModifiers.None;
-
+            bool seenNoDuplicates = true;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs]: Consider naming those positively (`seenDuplicates`, `seenAccessibilityDuplicates`).

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs]: I'm preserving the logic as it was in ParseModifiers.  I think that makes the review a bit easier.

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: #fixed

File: src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs [hunk 4]
```diff
@@ -44,7 +45,8 @@ internal static class ModifierUtils
                 switch (oneError)
                 {
                     case DeclarationModifiers.Partial:
-                        diagnostics.Add(ErrorCode.ERR_PartialMethodOnlyMethods, errorLocation);
+                        // Errors about 'partial' are reported in ToDeclarationModifiers.  So no need to report
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs]: The comment is not quite right. There is one path that calls `CheckModifiers` directly. See `SourceMemberContainerSymbol.cs` `MakeAndCheckTypeModifiers`.

File: src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationParsingTests.cs [hunk 5]
```diff
@@ -5269,6 +5269,61 @@ class Program {
...
+                Diagnostic(ErrorCode.ERR_SemicolonExpected, "partial").WithLocation(5, 13),
+                // (6,13): error CS1525: Invalid expression term 'partial'
+                //     partial partial void PM()
+                Diagnostic(ErrorCode.ERR_InvalidExprTerm, "partial").WithArguments("partial").WithLocation(6, 13),
+                // (6,13): error CS1002: ; expected
+                //     partial partial void PM()
+                Diagnostic(ErrorCode.ERR_SemicolonExpected, "partial").WithLocation(6, 13),
+                // (6,13): error CS0102: The type 'PartialPartial' already contains a definition for ''
+                //     partial partial void PM()
+                Diagnostic(ErrorCode.ERR_DuplicateNameInClass, "").WithArguments("PartialPartial", "").WithLocation(6, 13),
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationParsingTests.cs]: The empty string seems strange. Is that intended?

File: src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs]: True.  Let me investigate.  We may have a test hole here.

File: src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationParsingTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationParsingTests.cs]: will do.

**Pull request merged**